File: 297. Serialize and Deserialize Binary Tree/main.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func main() {
	root := &TreeNode{
		Val: 1,
		Left: &TreeNode{
			Val: 2,
			Left: &TreeNode{
				Val: 3,
			},
			Right: &TreeNode{
				Val: 4,
				Left: &TreeNode{
					Val: 5,
					Left: &TreeNode{
						Val: 6,
					},
					Right: &TreeNode{
						Val: 7,
					},
				},
			},
		},
	}

	ser := Constructor()
	deser := Constructor()
	data := ser.serialize(root)
	ans := deser.deserialize(data)

	fmt.Println(ans)
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type Codec struct {
	splitStr string
	nilStr   string
}

func Constructor() Codec {
	return Codec{
		splitStr: ",",
		nilStr:   "#",
	}
}

// Serializes a tree to a single string.
func (this *Codec) serialize(root *TreeNode) string {
	if root == nil {
		return this.nilStr
	}

	res := strconv.Itoa(root.Val)
	left := this.serialize(root.Left)
	right := this.serialize(root.Right)

	res = res + this.splitStr + left + this.splitStr + right
	return res
}

// Deserializes your encoded data to tree.
func (this *Codec) deserialize(data string) *TreeNode {
	splitted := strings.Split(data, this.splitStr)

	node, _ := de(splitted, 0, this.nilStr)
	return node
}

func de(arr []string, rangeStart int, nilStr string) (*TreeNode, int) {
	if rangeStart > len(arr) {
		return nil, rangeStart
	}

	if arr[rangeStart] == nilStr {
		return nil, rangeStart
	}

	root, _ := strconv.Atoi(arr[rangeStart])

	node := &TreeNode{
		Val: root,
	}

	var end int
	node.Left, end = de(arr, rangeStart+1, nilStr)
	node.Right, end = de(arr, end+1, nilStr)

	return node, end
}

/**
 * Your Codec object will be instantiated and called as such:
 * ser := Constructor();
 * deser := Constructor();
 * data := ser.serialize(root);
 * ans := deser.deserialize(data);
 */

File: 297. Serialize and Deserialize Binary Tree/main.py
# Definition for a binary tree node.
from typing import List, Optional


class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if not root:
            return "n"

        left = self.serialize(root.left)
        right = self.serialize(root.right)

        return str(root.val) + "," + left + "," + right

        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        def de(data: List[str], rangeStart) -> Optional[TreeNode]:
            if rangeStart >= len(data):
                return None
            if data[rangeStart] == 'n':
                return None, rangeStart
            
            root = TreeNode(int(data[rangeStart]))
            root.left, end = de(data, rangeStart+1)
            root.right, end = de(data, end + 1)

            return root, end


        splitted = data.split(",")
        res, _ = de(splitted, 0)
        return res
        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))